### ISSUE TITLE
curl[http2]:x64-windows-static requires nghttp2.dll

### ISSUE THREAD

**Issue by paul-sorensen** [microsoft/vcpkg#11492]
**Describe the bug**
This is perhaps a duplicate of #8923, although the author of that issue said they eventually got it working while I've not been able to.. To summarize, it seems the problem is with a dependent package - nghttp2 - and not curl itself.

When curl is installed statically and includes http2 support via nghttp2, a program that is run will display an error dialog about not being able to find nghttp2.dll even though both were installed statically. The message in the error dialog is "The code execution cannot proceed because nghttp2.dll was not found. Reinstalling the program may fix this problem."

**Environment**
- OS: Windows 10
- Compiler: Microsoft (R) C/C++ Optimizing Compiler Version 19.26.28805 for x64

**To Reproduce**
1. ./vcpkg install curl[http2]:x64-windows-static. 
2. Include curl in an application
3. Run the application

**Additional context**
This occurs because the source code of nghttp2 expects a symbol `NGHTTP2_STATICLIB` defined so it won't decorate functions as DLL exports. Here is the relevant section of nghttp2.h:
```
#ifdef NGHTTP2_STATICLIB
#  define NGHTTP2_EXTERN
#elif defined(WIN32) || (__has_declspec_attribute(dllexport) &&                \
                         __has_declspec_attribute(dllimport))
#  ifdef BUILDING_NGHTTP2
#    define NGHTTP2_EXTERN __declspec(dllexport)
#  else /* !BUILDING_NGHTTP2 */
#    define NGHTTP2_EXTERN __declspec(dllimport)
#  endif /* !BUILDING_NGHTTP2 */
#else    /* !defined(WIN32) */
#  ifdef BUILDING_NGHTTP2
#    define NGHTTP2_EXTERN __attribute__((visibility("default")))
#  else /* !BUILDING_NGHTTP2 */
#    define NGHTTP2_EXTERN
#  endif /* !BUILDING_NGHTTP2 */
#endif   /* !defined(WIN32) */
```

As stated in #8923 the curl build log includes the correct symbols. Example from `vcpkg/buildtrees/curl/install-x64-windows-static-dbg-out.log`:
```
[1/144] C:\PROGRA~2\MICROS~4\2019\COMMUN~1\VC\Tools\MSVC\1426~1.288\bin\Hostx64\x64\cl.exe   -DBUILDING_LIBCURL -DCURLDEBUG -DCURL_HIDDEN_SYMBOLS -DDEBUGBUILD -DHAVE_CONFIG_H -DNGHTTP2_STATICLIB -D_CRT_NONSTDC_NO_DEPRECATE -D_CRT_SECURE_NO_DEPRECATE -D_USRDLL -D_WIN32_WINNT=0x0600 -ID:\Development\cpp\vcpkg\buildtrees\curl\src\url-7_68_0-cd669cc759\include -Ilib\..\include -ID:\Development\cpp\vcpkg\buildtrees\curl\src\url-7_68_0-cd669cc759\lib\.. -ID:\Development\cpp\vcpkg\buildtrees\curl\src\url-7_68_0-cd669cc759\lib\..\include -Ilib\.. -ID:\Development\cpp\vcpkg\buildtrees\curl\src\url-7_68_0-cd669cc759\lib -Ilib -ID:\Development\cpp\vcpkg\installed\x64-windows-static\include /DWIN32 /D_WINDOWS /W4 /utf-8 /MP  /D_DEBUG /MTd /Z7 /Ob0 /Od /RTC1 /showIncludes /Folib\CMakeFiles\libcurl.dir\timeval.c.obj /Fdlib\CMakeFiles\libcurl.dir\libcurl.pdb /FS -c D:\Development\cpp\vcpkg\buildtrees\curl\src\url-7_68_0-cd669cc759\lib\timeval.c
Microsoft (R) C/C++ Optimizing Compiler Version 19.26.28805 for x64
Copyright (C) Microsoft Corporation.  All rights reserved.
```

The above compilation command includes `-DNGHTTP2_STATICLIB`, so the curl build generates the correct calling convention for the nghttp2 library. However, when inspecting the nghttp2 build log, it does not have the same symbol definition. Example from `vcpkg/buildtrees/nghttp2/install-x64-windows-static-dbg-out.log`:
```
[1/25] C:\PROGRA~2\MICROS~4\2019\COMMUN~1\VC\Tools\MSVC\1426~1.288\bin\Hostx64\x64\cl.exe   -DBUILDING_NGHTTP2 -DHAVE_CONFIG_H -Dnghttp2_EXPORTS -I. -ID:\Development\cpp\vcpkg\buildtrees\nghttp2\src\v1.39.2-d44e9e13e6\lib\includes -Ilib\includes /DWIN32 /D_WINDOWS /W3 /utf-8 /MP  /D_DEBUG /MTd /Z7 /Ob0 /Od /RTC1 /showIncludes /Folib\CMakeFiles\nghttp2.dir\nghttp2_npn.c.obj /Fdlib\CMakeFiles\nghttp2.dir\ /FS -c D:\Development\cpp\vcpkg\buildtrees\nghttp2\src\v1.39.2-d44e9e13e6\lib\nghttp2_npn.c
Microsoft (R) C/C++ Optimizing Compiler Version 19.26.28805 for x64
Copyright (C) Microsoft Corporation.  All rights reserved.
```
So, the functions compiled into nghttp2.lib still think they are in DLL mode due to the missing symbol definition. Not that it makes a difference, but the same missing symbol occurs when nghttp2 is installed by itself and not as a dependency of curl.

**Comment by paul-sorensen** [comment]
I was able to get this working by editing two files:
1. In `vcpkg/ports/nghttps2/portfile.cmake` replace this text
```
vcpkg_configure_cmake(
    SOURCE_PATH ${SOURCE_PATH}
    PREFER_NINJA
    OPTIONS
        -DENABLE_LIB_ONLY=ON
        -DENABLE_ASIO_LIB=OFF
)
```
with this text
```
if(VCPKG_LIBRARY_LINKAGE STREQUAL static)
    vcpkg_configure_cmake(
        SOURCE_PATH ${SOURCE_PATH}
        PREFER_NINJA
        OPTIONS
            -DENABLE_LIB_ONLY=ON
            -DENABLE_ASIO_LIB=OFF
            -DNGHTTP2_STATICLIB=1
    )
else()
vcpkg_configure_cmake(
    SOURCE_PATH ${SOURCE_PATH}
    PREFER_NINJA
    OPTIONS
        -DENABLE_LIB_ONLY=ON
        -DENABLE_ASIO_LIB=OFF
)
endif()
```
this will configure cmake with the correct parameter. However, CMakeLists.txt for nghttp2 also needs to recognize it. In `vcpkg/buildtrees/nghttp2/src/.../CMakeLists.txt` add the following:
```
if(NGHTTP2_STATICLIB)
    add_definitions(-DNGHTTP2_STATICLIB)
endif()
```
I don't know if it matters where this is added; I just added it near where another definition was added. I am most certainly not a vcpkg or cmake expert so there is probably a better way to do this. But after reinstalling nghttp2 with these changes I can see in the compilation log that `-DNGHTTP2_STATICLIB` is included, and to confirm, my program now starts without prompting for the missing DLL.

**Comment by thothothotho** [comment]
I can confirm that the problem also happens on linux.  The default triplet builds everything static.
But looking into `vcpkg/installed/x64-linux/lib/` shows that only libnghttp2.so* is  built.

**Comment by dnikolic-zfx** [comment]
I've ran into the same issue, and I'm able to reproduce it on windows after #11638 was merged.

Running the following command still produces dynamic .lib file on windows:
`.\vcpkg.exe install nghttp2:x86-windows-static`

Running the application linked with the nghttp2.lib built this way prompts for nghttp2.dll even tho it is supposed to be a static lib.
